gobblin-core/src/main/java/gobblin/source/extractor/extract/sftp/SftpLightWeightFileSystem.java

```diff
@@ -0,0 +1,270 @@
+      for (String name : fileNames) {
+        Path filePath = new Path(name);
+        if (VALID_PATH_FILTER.accept(filePath)) {
+          status.add(getFileStatus(new Path(path, filePath)));
+        }
+      }
+      return status.toArray(new FileStatus[status.size()]);
+    } catch (FileBasedHelperException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public boolean mkdirs(Path path, FsPermission permission) throws IOException {
+    try {
+      ChannelSftp channel = fsHelper.getSftpChannel();
+      channel.mkdir(path.toString());
+      channel.chmod((int) permission.toShort(), path.toString());
+      channel.disconnect();
+    } catch (SftpException e) {
+      throw new IOException(e);
+    }
+    return true;
+  }
+
+  @Override
+  public FSDataInputStream open(Path path, int bufferSize) throws IOException {
+    SftpGetMonitor monitor = new SftpGetMonitor();
+    try {
+      ChannelSftp channelSftp = fsHelper.getSftpChannel();
+      InputStream is = channelSftp.get(path.toString(), monitor);
+      return new FSDataInputStream(new BufferedFSInputStream(new SftpFsFileInputStream(is, channelSftp), bufferSize));
+    } catch (SftpException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public FSDataInputStream open(Path path) throws IOException {
+    return open(path, DEFAULT_BUFFER_SIZE);
+  }
+
+  @Override
+  public boolean rename(Path oldpath, Path newpath) throws IOException {
+    try {
+      ChannelSftp channelSftp = fsHelper.getSftpChannel();
+      channelSftp.rename(oldpath.toString(), newpath.toString());
+      channelSftp.disconnect();
+    } catch (SftpException e) {
+      throw new IOException(e);
+    }
+    return true;
+  }
+
+  @Override
+  public void setWorkingDirectory(Path path) {
+    try {
+      ChannelSftp channelSftp = fsHelper.getSftpChannel();
+      channelSftp.lcd(path.toString());
+      channelSftp.disconnect();
+    } catch (SftpException e) {
+      throw new RuntimeException("Failed to set working directory", e);
+    }
+  }
+
+  @Override
+  public void close() {
+    fsHelper.close();
+  }
+
+  @Override
+  public FSDataOutputStream append(Path arg0, int arg1, Progressable arg2) throws IOException {
+    throw new UnsupportedOperationException("Not implemented");
+  }
+
+  @Override
+  public FSDataOutputStream create(Path arg0, FsPermission arg1, boolean arg2, int arg3, short arg4, long arg5,
+      Progressable arg6) throws IOException {
+    throw new UnsupportedOperationException("Not implemented");
+  }
+
+  /**
+   * A {@link SeekableFSInputStream} that holds a handle on the Sftp {@link Channel} used to open the
+   * {@link InputStream}. The {@link Channel} is disconnected when {@link InputStream#close()} is called.
+   */
+  private class SftpFsFileInputStream extends SeekableFSInputStream {
+
+    private Channel channel;
+
+    public SftpFsFileInputStream(InputStream in) {
```

Comment:
why is this needed? It will cause the close() method to throw an NPE for this.channel.